internal/engine/selectors/selectors.go

```diff
@@ -0,0 +1,372 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate go run go.uber.org/mock/mockgen -package mock_$GOPACKAGE -destination=./mock/$GOFILE -source=./$GOFILE
+
+// Package selectors provides utilities for selecting entities based on profiles using CEL
+package selectors
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/stacklok/minder/internal/engine/entities"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// celEnvFactory is an interface for creating CEL environments
+// for an entity. Each entity must implement this interface to be
+// usable in selectors
+type celEnvFactory interface {
+	createEnv() (*cel.Env, error)
+}
+
+// genericEnvFactory is a factory for creating a CEL environment
+// for the generic SelectorEntity type
+type genericEnvFactory struct{}
+
+func (_ *genericEnvFactory) createEnv() (*cel.Env, error) {
+	return newEnvForEntity(
+		"entity",
+		&minderv1.SelectorEntity{},
+		"minder.v1.SelectorEntity")
+}
+
+// repoEnvFactory is a factory for creating a CEL environment
+// for the SelectorRepository type representing a repository
+type repoEnvFactory struct{}
+
+func (_ *repoEnvFactory) createEnv() (*cel.Env, error) {
+	return newEnvForEntity(
+		"repository",
+		&minderv1.SelectorEntity_SelectorRepository{},
+		"minder.v1.SelectorEntity.SelectorRepository")
+}
+
+// artifactEnvFactory is a factory for creating a CEL environment
+// for the SelectorArtifact type representing an artifact
+type artifactEnvFactory struct{}
+
+func (_ *artifactEnvFactory) createEnv() (*cel.Env, error) {
+	return newEnvForEntity(
+		"artifact",
+		&minderv1.SelectorEntity_SelectorArtifact{},
+		"minder.v1.SelectorEntity.SelectorArtifact")
+}
```

Comment:
**nit:** I tend to prefer function types to interfaces when they have a single method, which is usually the case with factories.

(part 1)

```suggestion
// celEnvFactory is a function type for creating CEL environments
// for an entity. Each entity must implement this interface to be
// usable in selectors
type celEnvFactory func() (*cel.Env, error)

// genericEnvFactory is a factory for creating a CEL environment
// for the generic SelectorEntity type
func genericEnvFactory() (*cel.Env, error) {
	return newEnvForEntity(
		"entity",
		&minderv1.SelectorEntity{},
		"minder.v1.SelectorEntity")
}

// repoEnvFactory is a factory for creating a CEL environment
// for the SelectorRepository type representing a repository
func repoEnvFactory() (*cel.Env, error) {
	return newEnvForEntity(
		"repository",
		&minderv1.SelectorEntity_SelectorRepository{},
		"minder.v1.SelectorEntity.SelectorRepository")
}

// artifactEnvFactory is a factory for creating a CEL environment
// for the SelectorArtifact type representing an artifact
func artifactEnvFactory() (*cel.Env, error) {
	return newEnvForEntity(
		"artifact",
		&minderv1.SelectorEntity_SelectorArtifact{},
		"minder.v1.SelectorEntity.SelectorArtifact")
}
```

Reply:
I think you're right that it would make the code tidier. Another thing I was considering was to have a single factory that would return both the environment and the converter just to make it easier to extend the code to handle a new entity.

Reply:
suggestion taken, thank you